packages/arb-bridge-eth/contracts/rollup/Node.sol

```diff
@@ -17,131 +17,146 @@
+    using SafeMath for uint256;
 
-    modifier onlyRollup() {
-        require(msg.sender == rollup, "ROLLUP_ONLY");
-        _;
+    /**
+     * @notice Update child properties
+     * @param number The child number to set
+     */
+    function childCreated(NodeProps storage self, uint256 number) internal {
+        if (self.firstChildBlock == 0) {
+            self.firstChildBlock = block.number;
+        }
+        self.latestChildNumber = number;
     }
 
     /**
-     * @notice Mark the given staker as staked on this node
-     * @param _rollup Initial value of rollup
+     * @notice Update the child confirmed deadline
+     * @param deadline The new deadline to set
+     */
+    function newChildConfirmDeadline(NodeProps storage self, uint256 deadline) internal {
+        self.noChildConfirmedBeforeBlock = deadline;
+    }
+
+    /**
+     * @notice Check whether the current block number has met or passed the node's deadline
+     */
+    function requirePastDeadline(NodeProps memory self) internal view {
+        require(block.number >= self.deadlineBlock, "BEFORE_DEADLINE");
+    }
+
+    /**
+     * @notice Check whether the current block number has met or passed deadline for children of this node to be confirmed
+     */
+    function requirePastChildConfirmDeadline(NodeProps memory self) internal view {
+        require(block.number >= self.noChildConfirmedBeforeBlock, "CHILD_TOO_RECENT");
+    }
+}
+
+
+/**
+ * @notice Utility functions for Node
+ */
+library NodeLib {
+    using SafeMath for uint256;
+    
+    /**
+     * @notice Initialise a Node
      * @param _stateHash Initial value of stateHash
      * @param _challengeHash Initial value of challengeHash
      * @param _confirmData Initial value of confirmData
      * @param _prev Initial value of prev
      * @param _deadlineBlock Initial value of deadlineBlock
      */
     function initialize(
-        address _rollup,
         bytes32 _stateHash,
         bytes32 _challengeHash,
         bytes32 _confirmData,
         uint256 _prev,
         uint256 _deadlineBlock
-    ) external override {
-        require(_rollup != address(0), "ROLLUP_ADDR");
-        require(rollup == address(0), "ALREADY_INIT");
-        rollup = _rollup;
-        stateHash = _stateHash;
-        challengeHash = _challengeHash;
-        confirmData = _confirmData;
-        prev = _prev;
-        deadlineBlock = _deadlineBlock;
-        noChildConfirmedBeforeBlock = _deadlineBlock;
-    }
-
-    /**
-     * @notice Destroy this node
-     */
-    function destroy() external override onlyRollup {
-        safeSelfDestruct(msg.sender);
+    ) internal pure returns (Node memory) {
+        NodeProps memory props;
+        props.stateHash = _stateHash;
+        props.challengeHash = _challengeHash;
+        props.confirmData = _confirmData;
+        props.prev = _prev;
+        props.deadlineBlock = _deadlineBlock;
+        props.noChildConfirmedBeforeBlock = _deadlineBlock;
+
+        Node memory node;
```

Comment:
why do we return this empty memory node? seems error prone. we can prob just drop the return value here

Reply:
It has the props on it - we store the struct returned from this function later on